packages/dds/sequence/src/intervalCollection.ts

```diff
@@ -1156,6 +1161,200 @@ class EndpointIndex<TInterval extends ISerializableInterval> implements Interval
 	}
 }
 
+/**
+ * Collection of intervals.
+ *
+ * Provide additional APIs to support efficiently querying a collection of intervals whose endpoints fall within a specified range.
+ */
+export interface IEndpointInRangeIndex<TInterval extends ISerializableInterval>
+	extends IntervalIndex<TInterval> {
+	/**
+	 * @returns an array of all intervals contained in this collection whose endpoints locate in the range [start, end] (include both ends)
+	 */
+	findIntervalsWithEndpointInRange(start: number, end: number);
+}
+
+/**
+ * Collection of intervals.
+ *
+ * Provide additional APIs to support efficiently querying a collection of intervals whose startpoints fall within a specified range.
+ */
+export interface IStartpointInRangeIndex<TInterval extends ISerializableInterval>
+	extends IntervalIndex<TInterval> {
+	/**
+	 * @returns an array of all intervals contained in this collection whose startpoints locate in the range [start, end] (include both ends)
+	 */
+	findIntervalsWithStartpointInRange(start: number, end: number);
+}
+
+class EndpointInRangeIndex<TInterval extends ISerializableInterval>
+	implements IEndpointInRangeIndex<TInterval>
+{
+	private readonly intervalTree;
+	private static readonly alwaysCompareSmaller = Symbol();
+	private static readonly alwaysCompareLarger = Symbol();
+
+	constructor(
+		private readonly helpers: IIntervalHelpers<TInterval>,
+		private readonly client: Client,
+	) {
+		this.intervalTree = new RedBlackTree<TInterval, TInterval>((a: TInterval, b: TInterval) => {
+			const compareEndsResult = helpers.compareEnds(a, b);
+			if (compareEndsResult !== 0) {
+				return compareEndsResult;
+			}
+			/**
+			 * The Symbol is only applied on the transient interval, which is always intended to be the first input
+			 * in the comparison. Therefore, it is sufficient to only check the Symbol of 'a' in the comparison logic.
+			 */
+			if ((a as any)[EndpointInRangeIndex.alwaysCompareSmaller]) {
```

Comment:
rather than cast to `any` here and where you inject the symbols, I'd make a type local to this file and cast to that. this also gives a convenient place to write a comment about why this approach is necessary (and provide an example--the test case you showed me made the issue pretty clear)

Reply:
maybe convert the special symbols to module-level locals (rather than private statics; that'll minimize better anyway) and then do

```typescript
interface HasComparisonOverride {
    [alwaysCompareSmaller]: boolean;
    [alwaysCompareLarger]: boolean;
}
```

or analogous if you decide to convert to one symbol.